relayer/workers/ethrelayer/parachain-writer.go

```diff
@@ -95,41 +101,26 @@ func (wr *ParachainWriter) writeLoop(ctx context.Context) error {
 	for {
 		select {
 		case <-ctx.Done():
-			return wr.onDone(ctx)
-		case msgs, ok := <-wr.messages:
-			// check if channel is closed
+			return ctx.Err()
+		case payload, ok := <-wr.payloads:
 			if !ok {
 				return nil
 			}
 
-			var concreteMsgs []*chain.EthereumOutboundMessage
-			for _, msg := range msgs {
-				cmsg, ok := msg.(*chain.EthereumOutboundMessage)
-				if !ok {
-					return fmt.Errorf("Invalid message")
-				}
-				concreteMsgs = append(concreteMsgs, cmsg)
-			}
-
-			err := wr.WriteMessages(ctx, concreteMsgs)
+			err := wr.WriteHeader(ctx, payload.header)
 			if err != nil {
 				wr.log.WithFields(logrus.Fields{
-					"error": err,
-				}).Error("Failure submitting message to substrate")
+					"blockNumber": payload.header.HeaderData.(ethereum.Header).Number,
+					"error":       err,
+				}).Error("Failure submitting header to substrate")
 				return err
 			}
-		case header, ok := <-wr.headers:
-			// check if channel is closed
-			if !ok {
-				return nil
-			}
 
-			err := wr.WriteHeader(ctx, &header)
+			err = wr.WriteMessages(ctx, payload.messages)
```

Comment:
for a single payload, if WriteHeader succeeds but WriteMessages does not, will the relayer still recover or resume successfully?

Reply:
No, not right now. The plan is to sort that out as part of https://github.com/Snowfork/polkadot-ethereum/issues/350